src/coreclr/vm/jitinterface.cpp

```diff
@@ -10623,10 +10631,41 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTINTERFACE ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS_SPECIAL ||
-            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX)
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_ARRADDR_ST ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_LDELEMA_REF)
         {
             Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);
             _ASSERTE(pPrecode->GetType() == PRECODE_FIXUP);
+            
+            MethodDesc* helperMD = pPrecode->GetMethodDesc();
+            _ASSERT(helperMD != nullptr);
+
+            // Check if the target MethodDesc is already jitted to its final Tier
+            // so we no longer need to use indirections and can emit a direct call instead.
+            NativeCodeVersion::OptimizationTier tier;
+
+            // Avoid taking the lock for foreground jit compilations
+            if (TieredCompilationManager::IsBackgroundWorkerActive())
+            {
+                {
+                    CodeVersionManager::LockHolder codeVersioningLockHolder;
+                    NativeCodeVersion activeCodeVersion = helperMD->GetCodeVersionManager()->GetActiveILCodeVersion(helperMD)
+                        .GetActiveNativeCodeVersion(helperMD);
+                    tier = activeCodeVersion.GetOptimizationTier();
+                }
+
+                if (tier == NativeCodeVersion::OptimizationTier::OptimizationTier1 ||
+                    tier == NativeCodeVersion::OptimizationTier::OptimizationTierOptimized)
+                {
+                    finalTierAddr = (LPVOID)GetEEFuncEntryPoint(hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);
```

Comment:
I'm not familiar with this array, is `hlpDynamicFuncTable[dynamicFtnNum].pfnHelper` updated after tiering up? The native entry point for the active code version can be retrieved by `activeCodeVersion.GetNativeCode()`.

Reply:
GetNativeCode seems to be indeed the correct target, changed.